File: python/spiral.py

# Facebook Interview Sample
#
# 2D Spiral Array
#
# Find the pattern and complete the function:
# int[][] spiral(int n);
# where n is the size of the 2D array.
#
# Sample Result
#
# input = 3
# 123
# 894
# 765
#
# input = 4
# 01 02 03 04
# 12 13 14 05
# 11 16 15 06
# 10 09 08 07

def rotate_right(dx, dy):
    return (-dy, dx)


def make_2d_spiral(n):
    result = [[None for y in range(n)] for x in range(n)]

    def is_valid(x, y):
        return x >= 0 and x < n and y >= 0 and y < n and (result[y][x] is None)

    x = 0
    y = 0

    dx = 1
    dy = 0

    current_number = 1

    while True:
        result[y][x] = current_number
        current_number += 1

        if is_valid(x + dx, y + dy):
            x += dx
            y += dy
        else:
            (dx, dy) = rotate_right(dx, dy)

            if is_valid(x + dx, y + dy):
                x += dx
                y += dy

            else:
                return result


def print_2d_spiral(matrix):
    n = len(matrix)
    fmt = '{{:0{:d}d}}'.format(len(str(n * n)))

    result = ('\n'.join(
        ' '.join(
            fmt.format(matrix[y][x])
            for x in range(n)
        )
        for y in range(n)
    ))

    print(result)


def make_and_print_2d_spiral(n):
    print_2d_spiral(make_2d_spiral(n))

File: scala/samples/src/main/scala/spiral.scala
// Facebook Interview Sample

// 2D Spiral Array

// Find the pattern and complete the function:
// int[][] spiral(int n);
// where n is the size of the 2D array.

// Sample Result

// input = 3
// 123
// 894
// 765

// input = 4
// 01 02 03 04
// 12 13 14 05
// 11 16 15 06
// 10 09 08 07

import scala.annotation.tailrec

class Field private (val dimensions: Int, map: Map[(Int, Int), Int] = Map()):
  def set(x: Int, y: Int, n: Int) = Field(dimensions, map + ((x, y) -> n))

  def apply(x: Int, y: Int) = map.get((x, y))

  def isValid(x: Int, y: Int) =
    x >= 0 && x < dimensions && y >= 0 && y < dimensions

object Field:
  def empty(n: Int) = Field(n)

def rotateRight(dx: Int, dy: Int) = (-dy, dx)

def canSet(field: Field, x: Int, y: Int) =
  field.isValid(x, y) && field(x, y).isEmpty

def mk2dSpiral(n: Int) =

  @tailrec
  def go(
      field: Field,
      x: Int,
      y: Int,
      dx: Int,
      dy: Int,
      currentN: Int
  ): Field =
    val newField = field.set(x, y, currentN)

    if canSet(newField, x + dx, y + dy) then
      go(
        field = newField,
        x = x + dx,
        y = y + dy,
        dx = dx,
        dy = dy,
        currentN = currentN + 1
      )
    else
      val (newDx, newDy) = rotateRight(dx, dy)
      if canSet(newField, x + newDx, y + newDy) then
        go(
          field = newField,
          x = x + newDx,
          y = y + newDy,
          dx = newDx,
          dy = newDy,
          currentN = currentN + 1
        )
      else newField

  go(Field.empty(n), x = 0, y = 0, dx = 1, dy = 0, currentN = 1)

def fieldToString(field: Field) =
  val nDigits = (field.dimensions * field.dimensions).toString.length
  val fmt = s"%0${nDigits}d"

  (0 until field.dimensions)
    .map { y =>
      (0 until field.dimensions)
        .map { x =>
          fmt.format(field(x, y).getOrElse(0))
        }
        .mkString(" ")
    }
    .mkString("\n")

@main def print2dSpiral = println(
  fieldToString(
    mk2dSpiral(7)
  )
)
